<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.39">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>8&nbsp; Generalized Additive (Mixed) Models – NZILBB R, Stats, and Open Science Workshops</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../chapters/references.html" rel="next">
<link href="../chapters/qualtrics.html" rel="prev">
<link href="../global_images/fav.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-c83d448f4d53f2b7bca973f5b61b9f20.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-f3bff2d0a531157203fa60ae48b91af4.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../include/webex.css">
</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../chapters/others.html">Additional Topics</a></li><li class="breadcrumb-item"><a href="../chapters/gamms_1.html"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Generalized Additive (Mixed) Models</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="../index.html" class="sidebar-logo-link">
      <img src="../global_images/nzilbb-logo.svg" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
    <div class="sidebar-title mb-0 py-0">
      <a href="../">NZILBB R, Stats, and Open Science Workshops</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/nzilbb/statistics_workshops" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../chapters/foundations.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Foundations</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/getting_started.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Getting Started</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/data_processing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Data Processing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/exploratory_data_vis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Exploratory Data Visualisation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/linear_models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Linear Models</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/mixed_effects.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Mixed Effects</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../chapters/others.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Additional Topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/rstudio_server.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">RStudio Server</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/qualtrics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Loading Data from Qualtrics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/gamms_1.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Generalized Additive (Mixed) Models</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#overview" id="toc-overview" class="nav-link active" data-scroll-target="#overview"><span class="header-section-number">8.1</span> Overview</a></li>
  <li><a href="#introduction-to-gams" id="toc-introduction-to-gams" class="nav-link" data-scroll-target="#introduction-to-gams"><span class="header-section-number">8.2</span> Introduction to GAMs</a>
  <ul class="collapse">
  <li><a href="#why" id="toc-why" class="nav-link" data-scroll-target="#why"><span class="header-section-number">8.2.1</span> Why?</a></li>
  <li><a href="#what" id="toc-what" class="nav-link" data-scroll-target="#what"><span class="header-section-number">8.2.2</span> What?</a></li>
  <li><a href="#fitting-gams-with-mgcv" id="toc-fitting-gams-with-mgcv" class="nav-link" data-scroll-target="#fitting-gams-with-mgcv"><span class="header-section-number">8.2.3</span> Fitting GAMs with <code>mgcv</code></a></li>
  <li><a href="#model-diagnostics" id="toc-model-diagnostics" class="nav-link" data-scroll-target="#model-diagnostics"><span class="header-section-number">8.2.4</span> Model diagnostics</a></li>
  <li><a href="#plotting" id="toc-plotting" class="nav-link" data-scroll-target="#plotting"><span class="header-section-number">8.2.5</span> Plotting</a></li>
  <li><a href="#interactions" id="toc-interactions" class="nav-link" data-scroll-target="#interactions"><span class="header-section-number">8.2.6</span> Interactions</a></li>
  </ul></li>
  <li><a href="#the-parenthetical-m-mixed-effects" id="toc-the-parenthetical-m-mixed-effects" class="nav-link" data-scroll-target="#the-parenthetical-m-mixed-effects"><span class="header-section-number">8.3</span> The Parenthetical ‘M’: Mixed Effects</a>
  <ul class="collapse">
  <li><a href="#random-intercepts" id="toc-random-intercepts" class="nav-link" data-scroll-target="#random-intercepts"><span class="header-section-number">8.3.1</span> Random Intercepts</a></li>
  <li><a href="#random-smooths" id="toc-random-smooths" class="nav-link" data-scroll-target="#random-smooths"><span class="header-section-number">8.3.2</span> Random Smooths</a></li>
  </ul></li>
  <li><a href="#sec-auto" id="toc-sec-auto" class="nav-link" data-scroll-target="#sec-auto"><span class="header-section-number">8.4</span> Auto-correlation</a></li>
  <li><a href="#sec-hyp" id="toc-sec-hyp" class="nav-link" data-scroll-target="#sec-hyp"><span class="header-section-number">8.5</span> Hypothesis Testing</a>
  <ul class="collapse">
  <li><a href="#visual-methods" id="toc-visual-methods" class="nav-link" data-scroll-target="#visual-methods"><span class="header-section-number">8.5.1</span> Visual methods</a></li>
  <li><a href="#model-summary" id="toc-model-summary" class="nav-link" data-scroll-target="#model-summary"><span class="header-section-number">8.5.2</span> Model summary</a></li>
  <li><a href="#itsadugcompareml." id="toc-itsadugcompareml." class="nav-link" data-scroll-target="#itsadugcompareml."><span class="header-section-number">8.5.3</span> <code>itsadug::compareML()</code>.</a></li>
  </ul></li>
  <li><a href="#reporting-a-gamm" id="toc-reporting-a-gamm" class="nav-link" data-scroll-target="#reporting-a-gamm"><span class="header-section-number">8.6</span> Reporting a GAMM</a></li>
  <li><a href="#sec-resources" id="toc-sec-resources" class="nav-link" data-scroll-target="#sec-resources"><span class="header-section-number">8.7</span> Further Resources</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/nzilbb/statistics_workshops/edit/main/chapters/gamms_1.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/nzilbb/statistics_workshops/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../chapters/others.html">Additional Topics</a></li><li class="breadcrumb-item"><a href="../chapters/gamms_1.html"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Generalized Additive (Mixed) Models</span></a></li></ol></nav>
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Generalized Additive (Mixed) Models</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<!-- Switch from bam to gam for the small models. This is more defensible.-->
<section id="overview" class="level2" data-number="8.1">
<h2 data-number="8.1" class="anchored" data-anchor-id="overview"><span class="header-section-number">8.1</span> Overview</h2>
<p>This chapter is going to grow over four workshop sessions. The rough plan is:</p>
<ol type="1">
<li>Introduction to the idea of GAMs and how to specify parametric and smooth terms.</li>
<li>The parenthetical ‘M’: we’ll add random effects, looking at random intercepts, slopes, and smooths.</li>
<li>Auto-correlation and model comparison: we’ll consider various strategies for testing for and controlling autocorrelation and methods for comparing models.</li>
<li>By-factor smooths and reporting models: We’ll consider the varieties of by-factor smooths available and how to implement them, before turning to how to report GA(M)Ms in papers and across supplementary material.</li>
</ol>
<p>We will be using the following libraries:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The usual suspects</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(here)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co"># GAMM-specific libraries</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mgcv)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(itsadug)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gratia)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Non-essential. Used in my `bam` to determine how many CPU cores to use.</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(parallel)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="co"># NZILBB vowel package</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="co"># If you do not have this use the following lines of code:</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="co"># install.packages('remotes')</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="co"># remotes::install_github('nzilbb/nzilbb_vowels')</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(nzilbb.vowels)</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Set ggplot theme</span></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_bw</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This workshop is heavily indebted to the workshops put together by Márton Sóskuthy and Martijn Wieling. Links to these are provided in <a href="#sec-resources" class="quarto-xref"><span>Section 8.7</span></a>.</p>
</section>
<section id="introduction-to-gams" class="level2" data-number="8.2">
<h2 data-number="8.2" class="anchored" data-anchor-id="introduction-to-gams"><span class="header-section-number">8.2</span> Introduction to GAMs</h2>
<section id="why" class="level3" data-number="8.2.1">
<h3 data-number="8.2.1" class="anchored" data-anchor-id="why"><span class="header-section-number">8.2.1</span> Why?</h3>
<p>Straight lines have a lot of advantages. They can be completely specified by two numbers: how steep they are (the <strong>slope</strong>) and where they intersect the <span class="math inline">\(y\)</span>-axis (the <strong>intercept</strong>). Fitting a straight line through a collection of points is just a matter of finding the optimum slope and intercept.</p>
<p>But <strong>sometimes straight lines aren’t enough</strong>. There are plenty of effects in nature which do not follow a straight line. There are plenty of examples of trajectories with non-linear behaviour in the study of language. For instance, consider the following trajectory for the <span class="smallcaps">price</span> vowel from ONZE via <span class="citation" data-cites="soskuthyHorizontalDiphthongShift2019">(<a href="references.html#ref-soskuthyHorizontalDiphthongShift2019" role="doc-biblioref">Sóskuthy, Hay, and Brand 2019</a>)</span>.<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a></p>
<div class="cell">
<details class="code-fold">
<summary>To view the code click here</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Source: https://osf.io/74mza</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Load all data (we will use the full set later)</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>price <span class="ot">&lt;-</span> <span class="fu">read_rds</span>(<span class="fu">here</span>(<span class="st">'data'</span>, <span class="st">'price_anon.rds'</span>))</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="co"># The dataset will be explained in full below.</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Pull out a single trajectory.</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>price <span class="ot">&lt;-</span> price <span class="sc">|&gt;</span> </span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(id <span class="sc">==</span> <span class="st">"price_58"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">cols =</span> f1<span class="sc">:</span>f2,</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="st">"formant_type"</span>,</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">values_to =</span> <span class="st">"formant_value"</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot it</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>price_plot <span class="ot">&lt;-</span> price <span class="sc">|&gt;</span> </span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> time,</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> formant_value,</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour =</span> formant_type</span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a>    <span class="at">colour =</span> <span class="st">"Formant type"</span>,</span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">"Frequency (Hz)"</span>,</span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">"Time (s)"</span></span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a>price_plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="gamms_1_files/figure-html/price-plot-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption>F1 and F2 of a PRICE vowel.</figcaption>
</figure>
</div>
</div>
</div>
<p>If we try to fit this trajectory with straight lines, we get:</p>
<div class="cell">
<details class="code-fold">
<summary>To view the code click here</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>price_plot <span class="sc">+</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>, <span class="at">se =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="gamms_1_files/figure-html/price-plot-linear-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption>F1 and F2 of a PRICE vowel with linear model.</figcaption>
</figure>
</div>
</div>
</div>
<p>What we want instead, is a way to fit a non-linear relationship. GA(M)Ms provide one flexible way of doing this. A simple GAM for this trajectory looks like this:</p>
<div class="cell">
<details class="code-fold">
<summary>To view the code click here</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>price_plot <span class="sc">+</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"gam"</span>, <span class="at">se =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="gamms_1_files/figure-html/price-plot-gam-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption>F1 and F2 of a PRICE vowel with GAM model.</figcaption>
</figure>
</div>
</div>
</div>
<p>The same comments apply to trajectories taken from, e.g., tongue sensors or derived from video footage. They also apply at very different time scales. Consider GAMMs fit through full monologues <span class="citation" data-cites="wilsonblackOverlookedEffectAmplitude2023">(<a href="references.html#ref-wilsonblackOverlookedEffectAmplitude2023" role="doc-biblioref">Wilson Black et al. 2023</a>)</span>, or to formant values across the history of a dialect <span class="citation" data-cites="brandSystematicCovariationMonophthongs2021">(<a href="references.html#ref-brandSystematicCovariationMonophthongs2021" role="doc-biblioref">Brand et al. 2021</a>)</span>.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
GAMMs and Polynomial Models
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>You may have used polynomial models in the past. These models add terms such as <span class="math inline">\(x^2\)</span> and <span class="math inline">\(x^3\)</span> to models (where previously, only <span class="math inline">\(x\)</span> was included). You can do a lot with polynomial models, but they have some negative properties. For a brief and accessible explanation of the shortcomings of polynomial models compared to GAMMs see the motivating example section from the following Gavin Simpson video: <a href="https://youtu.be/Ukfvd8akfco?si=COncckzAvpqvfIfj&amp;t=107" class="uri">https://youtu.be/Ukfvd8akfco?si=COncckzAvpqvfIfj&amp;t=107</a>. A key issue with polynomial models, especially when the <span class="math inline">\(n\)</span> in <span class="math inline">\(x^n\)</span> gets very large, is called <a href="https://en.wikipedia.org/wiki/Runge%27s_phenomenon">Runge’s Phenomenon</a>. At the edges of the range of the data, high order polynomials become very unstable.</p>
<p><span class="citation" data-cites="soskuthyGeneralisedAdditiveMixed2017">Sóskuthy (<a href="references.html#ref-soskuthyGeneralisedAdditiveMixed2017" role="doc-biblioref">2017</a>)</span> introduces polynomial models as, in effect, a simple kind of GAMM. This is different from thinking of GAMMs as an <em>alternative</em> to polynomial models. Both are defensible views, but you may get confused if you try to think them both at the same time!</p>
</div>
</div>
</div>
</section>
<section id="what" class="level3" data-number="8.2.2">
<h3 data-number="8.2.2" class="anchored" data-anchor-id="what"><span class="header-section-number">8.2.2</span> What?</h3>
<p>We’ll start with a bit of mathematics. A linear model looks like this:</p>
<p><span class="math display">\[ y = \beta_0 + \beta_1 x_1 + \beta_2 x_2 + \ldots + \beta_n x_n + \epsilon \]</span> where the <span class="math inline">\(\beta\)</span>’s are the model <strong>coefficients</strong>, the <span class="math inline">\(x\)</span>’s are the model predictors, and the <span class="math inline">\(\epsilon\)</span> is an error term. The <span class="math inline">\(\beta_0\)</span> term is the aforementioned <strong>intercept</strong> and the other <span class="math inline">\(\beta_n\)</span>’s are <strong>slopes</strong>. These <strong>slopes</strong> estimate the effect of the attached predictor. For any predictor, all we get is a single slope and the intercept — a straight line.</p>
<p>GAMs replace each of the <span class="math inline">\(\beta_n x_n\)</span> terms with a <strong>function</strong> of <span class="math inline">\(x_n\)</span>. This function will usually be some kind of <strong>smooth</strong>. We’ll look at this visually in a moment. But, mathematically, the overall GAM model looks like this:</p>
<p><span class="math display">\[  y = \beta_0 + f_1(x_1) + f_2(x_2) + \ldots + f_n(x_n) + \epsilon .\]</span> The <strong>parameters</strong> have been replaces by functions. Now, the nature of the relationship between a given predictor and <span class="math inline">\(y\)</span> need not be a straight line.</p>
<p>The functions we choose attempt to balance <strong>smoothness</strong> and <strong>wiggliness</strong>. Wiggliness simply means deviation from a straight line. Smoothness indicates a lack of bumps. We are speaking intuitively, but this language is used by the mathematicians as well!.</p>
<p>Let’s consider one class of <strong>smooth</strong>, and how it can be used to balance these two demands: splines. Here, a set of <strong>basis functions</strong> is summed together to create a smooth line through the data.</p>
<p>The basis functions for one common set of splines looks like this ( borrowing code from <a href="https://github.com/gavinsimpson/intro-gam-webinar-2020/blob/master/gam-intro.Rmd">Gavin Simpson</a>):</p>
<div class="cell">
<details class="code-fold">
<summary>To view the code click here</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Simulate 500 observations</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>N <span class="ot">&lt;-</span> <span class="dv">500</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">runif</span>(N),</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">ytrue =</span> <span class="fu">map_dbl</span>(</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>    x, </span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>    \(x) {x<span class="sc">^</span><span class="dv">11</span> <span class="sc">*</span> (<span class="dv">10</span> <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">-</span> x))<span class="sc">^</span><span class="dv">6</span> <span class="sc">+</span> ((<span class="dv">10</span> <span class="sc">*</span> (<span class="dv">10</span> <span class="sc">*</span> x)<span class="sc">^</span><span class="dv">3</span>) <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">-</span> x)<span class="sc">^</span><span class="dv">10</span>)}</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">ycent =</span> ytrue <span class="sc">-</span> <span class="fu">mean</span>(ytrue),</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">yobs  =</span> ycent <span class="sc">+</span> <span class="fu">rnorm</span>(N, <span class="at">sd =</span> <span class="fl">0.5</span>)</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>k <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>knots <span class="ot">&lt;-</span> <span class="fu">with</span>(data, <span class="fu">list</span>(<span class="at">x =</span> <span class="fu">seq</span>(<span class="fu">min</span>(x), <span class="fu">max</span>(x), <span class="at">length =</span> k)))</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>sm <span class="ot">&lt;-</span> <span class="fu">smoothCon</span>(<span class="fu">s</span>(x, <span class="at">k =</span> k, <span class="at">bs =</span> <span class="st">"cr"</span>), <span class="at">data =</span> data, <span class="at">knots =</span> knots)[[<span class="dv">1</span>]]<span class="sc">$</span>X</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(sm) <span class="ot">&lt;-</span> levs <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"f"</span>, <span class="fu">seq_len</span>(k))</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>basis <span class="ot">&lt;-</span> <span class="fu">pivot_longer</span>(<span class="fu">cbind</span>(sm, data), <span class="sc">-</span>(x<span class="sc">:</span>yobs), <span class="at">names_to =</span> <span class="st">'bf'</span>)</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>basis <span class="sc">|&gt;</span> </span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> x, </span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> value, </span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour =</span> bf</span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">colour =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">'x'</span>, <span class="at">y =</span> <span class="st">'b(x)'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-basis-functions" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-basis-functions-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="gamms_1_files/figure-html/fig-basis-functions-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-basis-functions-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;8.1: Cubic regression spline basis functions with 10 knots.
</figcaption>
</figure>
</div>
</div>
</div>
<p>Each distinct ‘basis function’ is in a different colour. We fit out actual data by multiplying the functions by appropriate coefficients (to change how <strong>high</strong> they are on the graph) and adding them together.</p>
<p>There are a number of <strong>knots</strong> in <a href="#fig-basis-functions" class="quarto-xref">Figure&nbsp;<span>8.1</span></a>. These are the points at which the functions a joined together (informally speaking) and at which we aim to ensure <strong>smoothness</strong>. For this set of basis functions, the knots are at (red points):</p>
<div class="cell">
<details class="code-fold">
<summary>To view the code click here</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Simulate 500 observations</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>N <span class="ot">&lt;-</span> <span class="dv">500</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">runif</span>(N),</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">ytrue =</span> <span class="fu">map_dbl</span>(</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>    x, </span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>    \(x) {x<span class="sc">^</span><span class="dv">11</span> <span class="sc">*</span> (<span class="dv">10</span> <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">-</span> x))<span class="sc">^</span><span class="dv">6</span> <span class="sc">+</span> ((<span class="dv">10</span> <span class="sc">*</span> (<span class="dv">10</span> <span class="sc">*</span> x)<span class="sc">^</span><span class="dv">3</span>) <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">-</span> x)<span class="sc">^</span><span class="dv">10</span>)}</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">ycent =</span> ytrue <span class="sc">-</span> <span class="fu">mean</span>(ytrue),</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">yobs  =</span> ycent <span class="sc">+</span> <span class="fu">rnorm</span>(N, <span class="at">sd =</span> <span class="fl">0.5</span>)</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>k <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>knots <span class="ot">&lt;-</span> <span class="fu">with</span>(data, <span class="fu">list</span>(<span class="at">x =</span> <span class="fu">seq</span>(<span class="fu">min</span>(x), <span class="fu">max</span>(x), <span class="at">length =</span> k)))</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>sm <span class="ot">&lt;-</span> <span class="fu">smoothCon</span>(<span class="fu">s</span>(x, <span class="at">k =</span> k, <span class="at">bs =</span> <span class="st">"cr"</span>), <span class="at">data =</span> data, <span class="at">knots =</span> knots)[[<span class="dv">1</span>]]<span class="sc">$</span>X</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(sm) <span class="ot">&lt;-</span> levs <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"f"</span>, <span class="fu">seq_len</span>(k))</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>basis <span class="ot">&lt;-</span> <span class="fu">pivot_longer</span>(<span class="fu">cbind</span>(sm, data), <span class="sc">-</span>(x<span class="sc">:</span>yobs), <span class="at">names_to =</span> <span class="st">'bf'</span>)</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>basis <span class="sc">|&gt;</span> </span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(</span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> x, </span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> value, </span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour =</span> bf</span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb6-28"><a href="#cb6-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(</span>
<span id="cb6-29"><a href="#cb6-29" aria-hidden="true" tabindex="-1"></a>    <span class="at">inherit.aes =</span> <span class="cn">FALSE</span>, </span>
<span id="cb6-30"><a href="#cb6-30" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb6-31"><a href="#cb6-31" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> x</span>
<span id="cb6-32"><a href="#cb6-32" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb6-33"><a href="#cb6-33" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="dv">0</span>,</span>
<span id="cb6-34"><a href="#cb6-34" aria-hidden="true" tabindex="-1"></a>    <span class="at">colour =</span> <span class="st">"red"</span>,</span>
<span id="cb6-35"><a href="#cb6-35" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> <span class="fu">as_tibble</span>(knots),</span>
<span id="cb6-36"><a href="#cb6-36" aria-hidden="true" tabindex="-1"></a>    <span class="at">size =</span> <span class="dv">5</span></span>
<span id="cb6-37"><a href="#cb6-37" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb6-38"><a href="#cb6-38" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">colour =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb6-39"><a href="#cb6-39" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">'x'</span>, <span class="at">y =</span> <span class="st">'b(x)'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-basis-functions-knots" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-basis-functions-knots-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="gamms_1_files/figure-html/fig-basis-functions-knots-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-basis-functions-knots-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;8.2: Cubic regression spline basis functions with 10 knots (red dots).
</figcaption>
</figure>
</div>
</div>
</div>
<p>Again, borrowing code from Gavin Simpson, we can see what this looks like for our simulated data.</p>
<div class="cell">
<details class="code-fold">
<summary>To view the code click here</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>beta <span class="ot">&lt;-</span> <span class="fu">coef</span>(<span class="fu">lm</span>(ycent <span class="sc">~</span> sm <span class="sc">-</span> <span class="dv">1</span>, <span class="at">data =</span> data))</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>wtbasis <span class="ot">&lt;-</span> <span class="fu">sweep</span>(sm, <span class="dv">2</span>L, beta, <span class="at">FUN =</span> <span class="st">"*"</span>)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(wtbasis) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(sm) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"F"</span>, <span class="fu">seq_len</span>(k))</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="do">## create stacked unweighted and weighted basis</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>basis <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(wtbasis) <span class="sc">%&gt;%</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> data<span class="sc">$</span>x,</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">spline_fit =</span> <span class="fu">pmap_dbl</span>(</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Yikes, bad coding here by me (JWB)</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>      <span class="fu">list</span>(F1, F2, F3, F4, F5, F6, F7, F8, F9, F10),</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>      sum</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>  ) </span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>basis_long <span class="ot">&lt;-</span> basis <span class="sc">|&gt;</span> </span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">cols =</span> <span class="fu">contains</span>(<span class="st">'F'</span>),</span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">values_to =</span> <span class="st">"value"</span>,</span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="st">"bf"</span></span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a>data <span class="sc">|&gt;</span> </span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(</span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> x,</span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> yobs</span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb7-30"><a href="#cb7-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.4</span>) <span class="sc">+</span></span>
<span id="cb7-31"><a href="#cb7-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(</span>
<span id="cb7-32"><a href="#cb7-32" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb7-33"><a href="#cb7-33" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> x,</span>
<span id="cb7-34"><a href="#cb7-34" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> value,</span>
<span id="cb7-35"><a href="#cb7-35" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour =</span> bf</span>
<span id="cb7-36"><a href="#cb7-36" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb7-37"><a href="#cb7-37" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> basis_long,</span>
<span id="cb7-38"><a href="#cb7-38" aria-hidden="true" tabindex="-1"></a>    <span class="at">inherit.aes =</span> <span class="cn">FALSE</span>,</span>
<span id="cb7-39"><a href="#cb7-39" aria-hidden="true" tabindex="-1"></a>    <span class="at">linewidth =</span> <span class="dv">1</span></span>
<span id="cb7-40"><a href="#cb7-40" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb7-41"><a href="#cb7-41" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(</span>
<span id="cb7-42"><a href="#cb7-42" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb7-43"><a href="#cb7-43" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> x,</span>
<span id="cb7-44"><a href="#cb7-44" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> spline_fit</span>
<span id="cb7-45"><a href="#cb7-45" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb7-46"><a href="#cb7-46" aria-hidden="true" tabindex="-1"></a>    <span class="at">inherit.aes =</span> <span class="cn">FALSE</span>,</span>
<span id="cb7-47"><a href="#cb7-47" aria-hidden="true" tabindex="-1"></a>    <span class="at">colour =</span> <span class="st">"black"</span>,</span>
<span id="cb7-48"><a href="#cb7-48" aria-hidden="true" tabindex="-1"></a>    <span class="at">linewidth =</span> <span class="dv">1</span>,</span>
<span id="cb7-49"><a href="#cb7-49" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> basis</span>
<span id="cb7-50"><a href="#cb7-50" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span> </span>
<span id="cb7-51"><a href="#cb7-51" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">colour =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-fit-spline" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-fit-spline-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="gamms_1_files/figure-html/fig-fit-spline-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-fit-spline-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;8.3: Simulated data (black dots) with the basis functions after multiplication by their weights (colourful lines) and the sum of the basis functions (black line).
</figcaption>
</figure>
</div>
</div>
</div>
<p>Spend some time looking at <a href="#fig-fit-spline" class="quarto-xref">Figure&nbsp;<span>8.3</span></a>. Convince yourself that if you added together the colourful lines you would get the black line. The easiest way to do this is to work one point on the <span class="math inline">\(x\)</span>-axis at a time. The case where <span class="math inline">\(x=0\)</span> is the easiest, where the only colourful line is the red one and it is at the same point on the <span class="math inline">\(y\)</span>-axis as the black line.</p>
<p>What about this <strong>wiggliness</strong> and <strong>smoothness</strong> trade off? We’ve already seen one way in which wiggliness can be controlled: the number of knots sets an upper limit on how wiggly the resulting smooth function can be. If we only had 3 knots, this is what we would get:</p>
<div class="cell">
<details class="code-fold">
<summary>To view the code click here</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>k <span class="ot">&lt;-</span> <span class="dv">3</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>knots <span class="ot">&lt;-</span> <span class="fu">with</span>(data, <span class="fu">list</span>(<span class="at">x =</span> <span class="fu">seq</span>(<span class="fu">min</span>(x), <span class="fu">max</span>(x), <span class="at">length =</span> k)))</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>sm <span class="ot">&lt;-</span> <span class="fu">smoothCon</span>(<span class="fu">s</span>(x, <span class="at">k =</span> k, <span class="at">bs =</span> <span class="st">"cr"</span>), <span class="at">data =</span> data, <span class="at">knots =</span> knots)[[<span class="dv">1</span>]]<span class="sc">$</span>X</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(sm) <span class="ot">&lt;-</span> levs <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"f"</span>, <span class="fu">seq_len</span>(k))</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>beta <span class="ot">&lt;-</span> <span class="fu">coef</span>(<span class="fu">lm</span>(ycent <span class="sc">~</span> sm <span class="sc">-</span> <span class="dv">1</span>, <span class="at">data =</span> data))</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>wtbasis <span class="ot">&lt;-</span> <span class="fu">sweep</span>(sm, <span class="dv">2</span>L, beta, <span class="at">FUN =</span> <span class="st">"*"</span>)</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(wtbasis) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(sm) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"F"</span>, <span class="fu">seq_len</span>(k))</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="do">## create stacked unweighted and weighted basis</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>basis <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(wtbasis) <span class="sc">%&gt;%</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> data<span class="sc">$</span>x,</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">spline_fit =</span> <span class="fu">pmap_dbl</span>(</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Yikes, bad coding here by me (JWB)</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>      <span class="fu">list</span>(F1, F2, F3),</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>      sum</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>  ) </span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>basis_long <span class="ot">&lt;-</span> basis <span class="sc">|&gt;</span> </span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">cols =</span> <span class="fu">contains</span>(<span class="st">'F'</span>),</span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">values_to =</span> <span class="st">"value"</span>,</span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="st">"bf"</span></span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a>data <span class="sc">|&gt;</span> </span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(</span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb8-29"><a href="#cb8-29" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> x,</span>
<span id="cb8-30"><a href="#cb8-30" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> yobs</span>
<span id="cb8-31"><a href="#cb8-31" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb8-32"><a href="#cb8-32" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb8-33"><a href="#cb8-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.4</span>) <span class="sc">+</span></span>
<span id="cb8-34"><a href="#cb8-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(</span>
<span id="cb8-35"><a href="#cb8-35" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb8-36"><a href="#cb8-36" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> x,</span>
<span id="cb8-37"><a href="#cb8-37" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> value,</span>
<span id="cb8-38"><a href="#cb8-38" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour =</span> bf</span>
<span id="cb8-39"><a href="#cb8-39" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb8-40"><a href="#cb8-40" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> basis_long,</span>
<span id="cb8-41"><a href="#cb8-41" aria-hidden="true" tabindex="-1"></a>    <span class="at">inherit.aes =</span> <span class="cn">FALSE</span>,</span>
<span id="cb8-42"><a href="#cb8-42" aria-hidden="true" tabindex="-1"></a>    <span class="at">linewidth =</span> <span class="dv">1</span></span>
<span id="cb8-43"><a href="#cb8-43" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb8-44"><a href="#cb8-44" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(</span>
<span id="cb8-45"><a href="#cb8-45" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb8-46"><a href="#cb8-46" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> x,</span>
<span id="cb8-47"><a href="#cb8-47" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> spline_fit</span>
<span id="cb8-48"><a href="#cb8-48" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb8-49"><a href="#cb8-49" aria-hidden="true" tabindex="-1"></a>    <span class="at">inherit.aes =</span> <span class="cn">FALSE</span>,</span>
<span id="cb8-50"><a href="#cb8-50" aria-hidden="true" tabindex="-1"></a>    <span class="at">colour =</span> <span class="st">"black"</span>,</span>
<span id="cb8-51"><a href="#cb8-51" aria-hidden="true" tabindex="-1"></a>    <span class="at">linewidth =</span> <span class="dv">1</span>,</span>
<span id="cb8-52"><a href="#cb8-52" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> basis</span>
<span id="cb8-53"><a href="#cb8-53" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span> </span>
<span id="cb8-54"><a href="#cb8-54" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">colour =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-fit-spline-bad" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-fit-spline-bad-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="gamms_1_files/figure-html/fig-fit-spline-bad-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-fit-spline-bad-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;8.4: Simulated data (black dots) with the basis functions after multiplication by their weights (colourful lines) and the sum of the basis functions (black line).
</figcaption>
</figure>
</div>
</div>
</div>
<p>The black line is our best possible fit to the data here, but it is no good. It needs to be wigglier.</p>
<p><strong>Knots</strong> are one determinant of wiggliness. But there is another: the <strong>smoothing parameter</strong>. This is used in order to penalise wiggliness when we fit a GAM model and is handled automatically by the <code>mgcv</code> package. In practice, it is determined <em>from the data</em>, rather than being manually specified. However, it is worth looking manually at what happens if we set the smoothing parameter too low and fail to sufficiently penalise wiggliness.</p>
<p>Here’s what an excessively wiggly smooth function looks like with the New Zealand English <span class="smallcaps">dress</span> vowel in ONZE:</p>
<div class="cell">
<details class="code-fold">
<summary>To view the code click here</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>onze_vowels_full <span class="sc">|&gt;</span> </span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lobanov_2</span>() <span class="sc">|&gt;</span> </span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>    gender <span class="sc">==</span> <span class="st">"F"</span>,</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>    vowel <span class="sc">==</span> <span class="st">"DRESS"</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span> </span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(speaker) <span class="sc">|&gt;</span> </span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">F1_lob2 =</span> <span class="fu">mean</span>(F1_lob2),</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">yob =</span> <span class="fu">first</span>(yob)</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span> </span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> yob,</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> F1_lob2</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(</span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">method =</span> <span class="st">"gam"</span>, </span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">formula =</span>  y <span class="sc">~</span> <span class="fu">s</span>(x, <span class="at">bs =</span> <span class="st">"cs"</span>, <span class="at">k =</span> <span class="dv">50</span>, <span class="at">sp=</span><span class="fl">0.01</span>), </span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">se=</span><span class="cn">FALSE</span></span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-fit-spline-lowsmooth" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-fit-spline-lowsmooth-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="gamms_1_files/figure-html/fig-fit-spline-lowsmooth-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-fit-spline-lowsmooth-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;8.5: Mean normalised F1 for female speakers in ONZE by year of birth. Smoothing parameter set to 0.01.
</figcaption>
</figure>
</div>
</div>
</div>
<p>On the other side, we can set the smoothing parameter too high. If we do, we’ll end up with a straight line.</p>
<div class="cell">
<details class="code-fold">
<summary>To view the code click here</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>mean_onze_full <span class="ot">&lt;-</span> onze_vowels_full <span class="sc">|&gt;</span> </span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lobanov_2</span>() <span class="sc">|&gt;</span> </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>    vowel <span class="sc">==</span> <span class="st">"DRESS"</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span> </span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(speaker) <span class="sc">|&gt;</span> </span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">F1_lob2 =</span> <span class="fu">mean</span>(F1_lob2),</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">yob =</span> <span class="fu">first</span>(yob),</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">speech_rate =</span> <span class="fu">mean</span>(speech_rate),</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">gender =</span> <span class="fu">first</span>(gender)</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>mean_onze_full <span class="sc">|&gt;</span> </span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(gender <span class="sc">==</span> <span class="st">"F"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(</span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> yob,</span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> F1_lob2</span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(</span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">method =</span> <span class="st">"gam"</span>, </span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a>    <span class="at">formula =</span>  y <span class="sc">~</span> <span class="fu">s</span>(x, <span class="at">bs =</span> <span class="st">"cs"</span>, <span class="at">k =</span> <span class="dv">50</span>, <span class="at">sp=</span><span class="dv">10000000</span>), </span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a>    <span class="at">se=</span><span class="cn">FALSE</span></span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-fit-spline-highsmooth" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-fit-spline-highsmooth-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="gamms_1_files/figure-html/fig-fit-spline-highsmooth-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-fit-spline-highsmooth-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;8.6: Mean normalised F1 for female speakers in ONZE by year of birth. Smoothing parameter set to 10,000,000.
</figcaption>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
The Bias-Variance Trade Off
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>We are now in the broad area of a core data science concept: the ‘bias variance tradeoff’. That is, with any statistical learning method, we can introduce errors with the assumptions of our model (<em>bias</em>) and errors due to excessively following small fluctuations in our data (<em>variance</em>). But the less bias we include, the more we will be lead astray by noise and vice versa. It’s a <em>tradeoff</em>. In the case we are looking at now, reducing the smoothing parameter is equivalent to decreasing bias and increasing variance.</p>
<p>See <a href="https://en.wikipedia.org/wiki/Bias%E2%80%93variance_tradeoff">the Wikipedia page</a>.</p>
</div>
</div>
</div>
</section>
<section id="fitting-gams-with-mgcv" class="level3" data-number="8.2.3">
<h3 data-number="8.2.3" class="anchored" data-anchor-id="fitting-gams-with-mgcv"><span class="header-section-number">8.2.3</span> Fitting GAMs with <code>mgcv</code></h3>
<p>How do we specify GAMs with the <code>mgcv</code> package? Let’s start with model formulae.</p>
<p>The most obvious this is the construction of smooth terms. These use the <code>s</code> function. Let’s look at the ONZE data from <a href="#fig-fit-spline-lowsmooth" class="quarto-xref">Figure&nbsp;<span>8.5</span></a> and <a href="#fig-fit-spline-highsmooth" class="quarto-xref">Figure&nbsp;<span>8.6</span></a>. Here we want normalised first formant values to vary with year of birth in a non-linear way. We want a <strong>smooth</strong> on year of birth.</p>
<p>What are the column names here?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>mean_onze_full</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 481 × 5
   speaker  F1_lob2   yob speech_rate gender
   &lt;fct&gt;      &lt;dbl&gt; &lt;int&gt;       &lt;dbl&gt; &lt;fct&gt; 
 1 CC_f_007  -0.765  1982        5.76 F     
 2 CC_f_010  -0.554  1947        5.06 F     
 3 CC_f_020  -0.242  1936        5.04 F     
 4 CC_f_024  -0.760  1973        5.61 F     
 5 CC_f_025  -0.747  1949        4.63 F     
 6 CC_f_027  -0.962  1981        5.22 F     
 7 CC_f_033  -0.725  1953        4.82 F     
 8 CC_f_040  -0.731  1955        4.24 F     
 9 CC_f_051  -0.769  1955        4.67 F     
10 CC_f_052  -0.838  1942        5.61 F     
# ℹ 471 more rows</code></pre>
</div>
</div>
<p>The variable we want to <em>explain</em> is <code>F1_lob2</code>. This contains normalised mean first formant values for each speaker. We want to explain it using `yob’, which we can see from the tibble output, is an integer. We’ll look at incorporating the other variables later.</p>
<p>The simplest version of the formula is this: <code>F1_lob2 ~ s(yob)</code>. But this is usually bad practice — we should be more explicit! As a general principle, relying on defaults is dangerous as they can change under you, causing your code to have a different outcome.</p>
<p>The first area to be explicit is the <strong>knots</strong>. The default value for many smoothing splines is <span class="math inline">\(10\)</span> and this is almost always fine. But we should think about it each time we fit a GAMM. So, an improvement: <code>F1_lob2 ~ s(yob, k = 10)</code>.</p>
<p>The second argument to highlight is <code>bs</code>. This says what kind of basis functions we are using. The default, <code>tp</code>, or thin plate regression splines, are fine. Typically this choice won’t make a big different to you. But I will add more detail here soon. Regardless, it is good to make this explicit. So our final version of this formula: <code>F1_lob2 ~ s(yob, k = 10, bs = "tp")</code>.</p>
<p>What do we <em>do</em> with this formula? We will use the function <code>bam</code> to fit our first GAM.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>We could just as easily use the <code>gam</code> function, but <code>bam</code> is optimised for large datasets.</p>
</div>
</div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Parallel Processing on macOS
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The <code>mgcv</code> package requires the <code>openmp</code> library to take advantage of the parallel processing available through <code>bam()</code>. This is typically not a problem for Windows and Linux installations of R. On macos, it can be a problem.</p>
<p>If you get a warning about <code>openmp</code> not being available on macOS. I suggest you install the mandatory tools listed here: <a href="https://mac.r-project.org/tools/" class="uri">https://mac.r-project.org/tools/</a> and then follow the instructions at <a href="https://mac.r-project.org/openmp/" class="uri">https://mac.r-project.org/openmp/</a>.</p>
<p>This is not likely to be a pleasant experience, especially if you don’t like battling with the command line. UC and NZILBB researchers and students, feel free to message me on Rocket.Chat or via <a href="mailto:joshua.black@canterbury.ac.nz" class="email">joshua.black@canterbury.ac.nz</a> and I may be able to help.</p>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>onze_fit <span class="ot">&lt;-</span> <span class="fu">bam</span>(</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">formula =</span> F1_lob2 <span class="sc">~</span> <span class="fu">s</span>(yob, <span class="at">k =</span> <span class="dv">10</span>, <span class="at">bs =</span> <span class="st">"tp"</span>),</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> mean_onze_full</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We obtain a summary for this model using the <code>summary</code> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(onze_fit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Family: gaussian 
Link function: identity 

Formula:
F1_lob2 ~ s(yob, k = 10, bs = "tp")

Parametric coefficients:
             Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept) -0.590123   0.008102  -72.84   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Approximate significance of smooth terms:
         edf Ref.df     F p-value    
s(yob) 3.633  4.481 178.6  &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

R-sq.(adj) =  0.625   Deviance explained = 62.8%
fREML = -138.75  Scale est. = 0.031571  n = 481</code></pre>
</div>
</div>
<p>This summary has two primary sections. The <code>Parametric coefficients</code>, which indicate the non-smooth aspects of the model. In this case, the model fits an intercept term, which sets the overall height of the smooth function. The <code>Approximate significance of smooth terms</code> section indicates, as it says, the approximate significance of our smooths. This is the GAM equivalent of the coefficient for a variable in a linear model.</p>
<p>In this model, we only have <code>s(yob)</code> to look at. We see that it has an <code>edf</code> or ‘estimated degrees of freedom’ of 3.633. This is an indication of how wiggly the line is. If the esimated degrees of freedom are 1, it’s pretty much a straight line. We also see a <code>p-value</code> entry. This indicates whether the shape of the smooth is statistically significantly different from a flat line at the intercept value. In this case, unsurprisingly, it is distinct from a flat line.</p>
<p>But there are some problems here. First, we are merging male and female data together here. What if we want to fit a smooth for both male and female speakers? Here we can us the <code>by</code> argument to <code>s()</code> <em>and</em> add a parametric term for <code>gender</code>. This results in:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>onze_fit_gender <span class="ot">&lt;-</span> <span class="fu">bam</span>(</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">formula =</span> F1_lob2 <span class="sc">~</span> gender <span class="sc">+</span> <span class="fu">s</span>(yob, <span class="at">by =</span> gender, <span class="at">k =</span> <span class="dv">10</span>, <span class="at">bs =</span> <span class="st">"tp"</span>),</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> mean_onze_full</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(onze_fit_gender)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Family: gaussian 
Link function: identity 

Formula:
F1_lob2 ~ gender + s(yob, by = gender, k = 10, bs = "tp")

Parametric coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept) -0.64493    0.01146 -56.255  &lt; 2e-16 ***
genderM      0.10157    0.01565   6.492 2.13e-10 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Approximate significance of smooth terms:
                 edf Ref.df      F p-value    
s(yob):genderF 2.754  3.394  72.13  &lt;2e-16 ***
s(yob):genderM 1.000  1.000 588.35  &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

R-sq.(adj) =  0.658   Deviance explained = 66.2%
fREML = -156.83  Scale est. = 0.028773  n = 481</code></pre>
</div>
</div>
<p>Now we have <em>two</em> intercept terms, one for the female speakers and one for the male, both of which are significant. Just as in generalised linear models, the <code>genderM</code> parametric term gives a <em>difference</em> from the female intercept. This indicates that the first formant value is on average higher for male speakers. In the smooth terms section we now see <code>s(yob):genderF</code> and <code>s(yob):genderM</code>. We get independent p-values for each. What we do not get is a representation of the <em>difference</em> between the smooth for the female speakers and the smooth for the male speakers.<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a></p>
<p>Note that the smooth for the male speakers in effectively a straight line. We will see this in a moment when we visualise.</p>
<p>We will add one more thing to this model before turning to diagnostics and plotting. What if we want <em>two</em> smooths? We know that speech rate can affect formant values. We can add this as an additional smooth term as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>onze_fit_rate <span class="ot">&lt;-</span> <span class="fu">bam</span>(</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">formula =</span> F1_lob2 <span class="sc">~</span> gender <span class="sc">+</span> </span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">s</span>(yob, <span class="at">by =</span> gender, <span class="at">k =</span> <span class="dv">10</span>, <span class="at">bs =</span> <span class="st">"tp"</span>) <span class="sc">+</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">s</span>(speech_rate, <span class="at">k =</span> <span class="dv">10</span>, <span class="at">bs =</span> <span class="st">"tp"</span>),</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> mean_onze_full</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(onze_fit_rate)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Family: gaussian 
Link function: identity 

Formula:
F1_lob2 ~ gender + s(yob, by = gender, k = 10, bs = "tp") + s(speech_rate, 
    k = 10, bs = "tp")

Parametric coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept) -0.64505    0.01148 -56.176  &lt; 2e-16 ***
genderM      0.10180    0.01568   6.492 2.14e-10 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Approximate significance of smooth terms:
                 edf Ref.df       F p-value    
s(yob):genderF 2.754  3.393  64.128  &lt;2e-16 ***
s(yob):genderM 1.000  1.000 488.583  &lt;2e-16 ***
s(speech_rate) 1.000  1.000   0.082   0.775    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

R-sq.(adj) =  0.658   Deviance explained = 66.2%
fREML = -153.06  Scale est. = 0.028828  n = 481</code></pre>
</div>
</div>
<p>In this case we don’t see a significant difference as a result of speech rate. This may be because we are working with mean values for the formants. Once we can include random effects, and thus multiple values from a single speaker, this will change!</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p><span class="citation" data-cites="wielingAnalyzingDynamicPhonetic2018">Wieling (<a href="references.html#ref-wielingAnalyzingDynamicPhonetic2018" role="doc-biblioref">2018</a>)</span> provides an example of building up a model from the ground up, exploring many different possible structures and including the R code. It is a good first port of call for looking at additional possible structures.</p>
</div>
</div>
</section>
<section id="model-diagnostics" class="level3" data-number="8.2.4">
<h3 data-number="8.2.4" class="anchored" data-anchor-id="model-diagnostics"><span class="header-section-number">8.2.4</span> Model diagnostics</h3>
<p>The primary port of call for model diagnostics in <code>mgcv</code> is the <code>gam.check()</code> function. One of the outputs is a text output to help determine if <span class="math inline">\(k\)</span> is too low.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gam.check</span>(onze_fit_rate)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Method: fREML   Optimizer: perf newton
full convergence after 15 iterations.
Gradient range [-1.027586e-06,1.027607e-07]
(score -153.0627 &amp; scale 0.02882816).
Hessian positive definite, eigenvalue range [2.671181e-07,238.0032].
Model rank =  29 / 29 

Basis dimension (k) checking results. Low p-value (k-index&lt;1) may
indicate that k is too low, especially if edf is close to k'.

                 k'  edf k-index p-value  
s(yob):genderF 9.00 2.75    1.00   0.530  
s(yob):genderM 9.00 1.00    1.00   0.460  
s(speech_rate) 9.00 1.00    0.93   0.065 .
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>The output above has an entry for each smooth term in the model. <code>k'</code> indicate the knots the smooth function has available. This, as explained above, is an upper limit on wiggliness. It will usually be one less than the value of <code>k</code> given in the model formula. The value <code>edf</code> indicates the actual wiggliness of the smooth. To determine if there is evidence of insufficient <code>k</code>, check whether the <code>p-value</code> is low <em>and</em> the <code>edf</code> is close to <code>k'</code>. If so, consider increasing <code>k</code> in the model. In this case, <code>k</code> is plenty high enough.</p>
<p>Let’s see a case where this doesn’t work well.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>sim_fit <span class="ot">&lt;-</span> <span class="fu">bam</span>(</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">formula =</span> yobs <span class="sc">~</span> <span class="fu">s</span>(x, <span class="at">k =</span> <span class="dv">3</span>, <span class="at">bs =</span> <span class="st">"cr"</span>),</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> data</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a><span class="fu">gam.check</span>(sim_fit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Method: fREML   Optimizer: perf newton
full convergence after 11 iterations.
Gradient range [-2.519727e-09,2.506653e-09]
(score 1287.688 &amp; scale 9.969049).
Hessian positive definite, eigenvalue range [0.4972726,249.001].
Model rank =  3 / 3 

Basis dimension (k) checking results. Low p-value (k-index&lt;1) may
indicate that k is too low, especially if edf is close to k'.

     k' edf k-index p-value    
s(x)  2   2    0.03  &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>Here <code>k'</code> and <code>edf</code> are the same, and the <code>p-value</code> is very low. Something is going wrong here. In fact, the problem here is the same problem we saw in <a href="#fig-fit-spline-bad" class="quarto-xref">Figure&nbsp;<span>8.4</span></a>.</p>
<p>If you run <code>gam.check</code> you will also see some diagnostic plots. I have suppressed them in this document. The <code>gratia</code> package provides a nice wrapper for the <code>gam.check</code> visualisations (via the <code>appraise</code> function). It has the advantage of being a single plot in <code>ggplot</code> format which will take on any global changes to you ‘theme’. That is, it will produce output which matches your other plots.</p>
<p>Let’s look at the bad fit to our simulated data again:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">appraise</span>(sim_fit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="gamms_1_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>These plots tell us something about the <em>residuals</em>. The assumption of our model is that these residuals will be normally distributed. That is, we assume that the variation which is left behind by our model looks like random observations from a normal distribution.</p>
<p>These diagnostic plots are <strong>not</strong> like this. At the top left, we should see a straight line of black points following the red line. However, we see that at the tails, at extreme values of a predictor, we are not getting what we would expect from a normal distribution. The histogram tells a similar story. This should look like a nice(ish) bell curve. But the most extreme warning signs are the two plots on the right. These show the model predictions plotted against the actual values. There should be no obvious (non-linear) pattern in these plots.</p>
<p>We already know what to do in this case, we need to increase <span class="math inline">\(k\)</span>! If we do, this is what we get:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>sim_fit_highk <span class="ot">&lt;-</span> <span class="fu">bam</span>(</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">formula =</span> yobs <span class="sc">~</span> <span class="fu">s</span>(x, <span class="at">k =</span> <span class="dv">20</span>, <span class="at">bs =</span> <span class="st">"cr"</span>),</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> data</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a><span class="fu">appraise</span>(sim_fit_highk)  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="gamms_1_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Much better! And the check of <code>k</code> looks OK too:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gam.check</span>(sim_fit_highk)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Method: fREML   Optimizer: perf newton
full convergence after 5 iterations.
Gradient range [-1.018439e-06,1.032479e-06]
(score 434.2567 &amp; scale 0.2813964).
Hessian positive definite, eigenvalue range [8.433327,249.2893].
Model rank =  20 / 20 

Basis dimension (k) checking results. Low p-value (k-index&lt;1) may
indicate that k is too low, especially if edf is close to k'.

       k'  edf k-index p-value
s(x) 19.0 17.7     1.1    0.98</code></pre>
</div>
</div>
<p>Returning to our model of the ONZE data, let’s use <code>appraise</code> again:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">appraise</span>(onze_fit_rate)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="gamms_1_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>This looks basically fine. But we can see heavier tails that we would usually want at either end of the QQ plot and the histogram. This is quite common in vocalic data. One way to handle this is to assume that the residuals follow a t distribution instead (these have fatter tails than normal distributions). We can do this with the <code>family</code> argument to <code>bam</code>. If we do this, the plots look a bit better:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>onze_fit_rate <span class="ot">&lt;-</span> <span class="fu">bam</span>(</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">formula =</span> F1_lob2 <span class="sc">~</span> gender <span class="sc">+</span> </span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">s</span>(yob, <span class="at">by =</span> gender, <span class="at">k =</span> <span class="dv">10</span>, <span class="at">bs =</span> <span class="st">"tp"</span>) <span class="sc">+</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">s</span>(speech_rate, <span class="at">k =</span> <span class="dv">10</span>, <span class="at">bs =</span> <span class="st">"tp"</span>),</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> mean_onze_full,</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">family =</span> <span class="fu">scat</span>(<span class="at">link=</span><span class="st">"identity"</span>)</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a><span class="fu">appraise</span>(onze_fit_rate)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="gamms_1_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We will see a case below where a t-distribution functions better than a standard normal distribution with formant data. It is worth checking this case-by-case though.</p>
</section>
<section id="plotting" class="level3" data-number="8.2.5">
<h3 data-number="8.2.5" class="anchored" data-anchor-id="plotting"><span class="header-section-number">8.2.5</span> Plotting</h3>
<p>Plotting smooths can be done in at least three ways:</p>
<ol type="1">
<li>Using a prediction function to generate predictions from the model and then plot them yourself. The advantage is high flexibility in your plots.</li>
<li>Use the <code>plot_smooth()</code> and related functions from <code>itsadug</code>.</li>
<li>Use the GAM plotting functions from <code>gratia</code>.</li>
</ol>
<p>Let’s look at the <code>plot_smooth()</code> function. This has been used in a lot of projects at NZILBB.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_smooth</span>(</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> <span class="co"># the model,</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">view =</span> <span class="co"># the name of the variable you want to plot.</span></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">cond =</span> <span class="co"># a named list containing the values of other terms in the model. </span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># if not given you will get mean values.</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot_all =</span> <span class="co"># The name of any factors which for which you want all levels to be</span></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># plotted</span></span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">rug =</span> <span class="co"># display a 'rug' at the bottom of the plot to indicate where there are</span></span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># actual observations.</span></span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, using these to plot the ONZE model, we get:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_smooth</span>(</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> onze_fit_rate,</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">view =</span> <span class="st">"yob"</span>,</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot_all =</span> <span class="st">"gender"</span>,</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">rug =</span> <span class="cn">TRUE</span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Summary:
    * gender : factor; set to the value(s): F, M. 
    * yob : numeric predictor; with 30 values ranging from 1864.000000 to 1982.000000. 
    * speech_rate : numeric predictor; set to the value(s): 4.86001666666667. 
    * NOTE : No random effects in the model to cancel.
 </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="gamms_1_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Note that the values given for other predictors are given in console output.</p>
<p>Now change the <code>view</code> and <code>cond</code> arguments and see what happens.</p>
<p>The <code>draw()</code> function from <code>gratia</code> is very useful for plotting the ‘partial effects’ of each smooth term. Let’s look at an example and then consider what this means in more detail.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">draw</span>(onze_fit_rate)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="gamms_1_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The difference between ‘summed effects’, which you get in <code>plot_smooth()</code>, and the ‘partial effects’ produced by <code>draw()</code>, is that summer effects include intercept terms and values for other predictors. We’ve seen above that <code>plot_smooth()</code> produces an output message which states what the level of the other predictors has been set to. The <code>draw()</code> function, and the <code>mgcv</code> function <code>plot.gam()</code>, instead show the effect of the smooth terms <em>individually</em>. These will be centred around <span class="math inline">\(0\)</span>.</p>
</section>
<section id="interactions" class="level3" data-number="8.2.6">
<h3 data-number="8.2.6" class="anchored" data-anchor-id="interactions"><span class="header-section-number">8.2.6</span> Interactions</h3>
<p>Often we are interested in how two predictors interact to generate a response. We have already seen one version of this: <em>continuous by factor</em> interactions created using the <code>by</code> argument to <code>s()</code>. We determined the change over time in ONZE <em>independently for male and female</em> speakers.</p>
<p>There are also methods for including <em>continuous by continuous</em> interactions in GAMMs. The two main functions to consider are <code>ti()</code> and <code>te()</code>.</p>
<p>The <code>ti()</code> function produces a ‘tensor product interaction’. With the example data from ONZE we have been using there are only two predictors which can be naturally treated as continuous: year of birth and speech rate. In order to interact these predictors, we could add a <code>ti()</code> term. This is done by the following code block. For the purposes of this example we remove gender from the model, but we will add it in again in a moment. The use of <code>k=5</code> just makes the default behaviour for this kind of smooth explicit.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>onze_fit_int <span class="ot">&lt;-</span> <span class="fu">bam</span>(</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">formula =</span> F1_lob2 <span class="sc">~</span> </span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">s</span>(yob, <span class="at">k =</span> <span class="dv">10</span>, <span class="at">bs =</span> <span class="st">"tp"</span>) <span class="sc">+</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">s</span>(speech_rate, <span class="at">k =</span> <span class="dv">10</span>, <span class="at">bs =</span> <span class="st">"tp"</span>) <span class="sc">+</span></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ti</span>(yob, speech_rate, <span class="at">k=</span><span class="dv">5</span>, <span class="at">bs =</span> <span class="st">"tp"</span>),</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> mean_onze_full,</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">family =</span> <span class="fu">scat</span>(<span class="at">link=</span><span class="st">"identity"</span>)</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a><span class="fu">draw</span>(onze_fit_int)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="gamms_1_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We now get a heat map, which represents the interaction. In order to read the heat map, find the values of the predictors on the <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span> axis and then use the colour guide to determine what to add to the partial effects of the predictors. For instance, if we look in the bottom left of the heatmap we see a patch of blue. This indicates that early speakers who are slow tend to have lower F1 values for <span class="smallcaps">dress</span>. This is to say, they start from a higher position in the vowel space (although still low relative to contemporary New Zealand English speakers). The heatmap helpfully greys out areas in which there are no datapoints.</p>
<p>What does this look like in a model summary? Here is the relevant table:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(onze_fit_int)<span class="sc">$</span>s.table</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                         edf   Ref.df         F    p-value
s(yob)              3.975988 4.909228 74.411993 0.00000000
s(speech_rate)      2.696854 3.447972  1.896666 0.12778914
ti(yob,speech_rate) 5.684495 7.421081  1.863619 0.06482433</code></pre>
</div>
</div>
<p>There is a new line for the <code>ti()</code> term. We see that it does not achieve statistical significance at the <span class="math inline">\(0.05\)</span> level.<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a></p>
<p>The function <code>te()</code> specifies a ‘tensor product smooth’. This includes the individual effects of any included predictors <em>and</em> their interaction. This means the formula for the model no longer has individual terms for year of birth and speech rate. This looks very simple:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>onze_fit_te <span class="ot">&lt;-</span> <span class="fu">gam</span>(</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">formula =</span> F1_lob2 <span class="sc">~</span> </span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">te</span>(yob, speech_rate, <span class="at">k=</span><span class="dv">5</span>, <span class="at">bs =</span> <span class="st">"tp"</span>),</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> mean_onze_full,</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">family =</span> <span class="fu">scat</span>(<span class="at">link=</span><span class="st">"identity"</span>)</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a><span class="fu">draw</span>(onze_fit_te)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="gamms_1_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Now there’s only one smooth term, carrying all of the relationship between speech rate, year of birth, their interaction, and the normalised F1 of the <span class="smallcaps">dress</span> vowel. This is, as one might imagine, less useful for testing specific hypotheses about the role of, say, speech rate, or year of birth. Indeed, if we look in the model summary, we will see one term:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(onze_fit_te)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Family: Scaled t(5.187,0.141) 
Link function: identity 

Formula:
F1_lob2 ~ te(yob, speech_rate, k = 5, bs = "tp")

Parametric coefficients:
             Estimate Std. Error z value Pr(&gt;|z|)    
(Intercept) -0.592324   0.007391  -80.14   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Approximate significance of smooth terms:
                      edf Ref.df Chi.sq p-value    
te(yob,speech_rate) 10.94  13.68  974.5  &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

R-sq.(adj) =  0.627   Deviance explained = 51.7%
-REML = -149.4  Scale est. = 1         n = 481</code></pre>
</div>
</div>
<p>There is no obvious way to unpack this information.</p>
<p>One final thing to note is that we can add <code>by</code> arguments to <code>ti()</code> and <code>te()</code> just as we did for <code>s()</code>. So, for instance, we might fit a model which has an interaction for speech rate, year of birth, <em>and</em> gender, as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>onze_fit_gen_int <span class="ot">&lt;-</span> <span class="fu">gam</span>(</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">formula =</span> F1_lob2 <span class="sc">~</span> </span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">s</span>(yob, <span class="at">k =</span> <span class="dv">10</span>, <span class="at">by=</span>gender, <span class="at">bs =</span> <span class="st">"tp"</span>) <span class="sc">+</span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">s</span>(speech_rate, <span class="at">k =</span> <span class="dv">10</span>, <span class="at">bs =</span> <span class="st">"tp"</span>) <span class="sc">+</span></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ti</span>(yob, speech_rate, <span class="at">by=</span>gender, <span class="at">k=</span><span class="dv">5</span>, <span class="at">bs =</span> <span class="st">"tp"</span>),</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> mean_onze_full,</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">family =</span> <span class="fu">scat</span>(<span class="at">link=</span><span class="st">"identity"</span>)</span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a><span class="fu">draw</span>(onze_fit_gen_int)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="gamms_1_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Now we have, in effect, a three way interaction. We get two distinct heatmaps, one for the interaction between year of birth and speech rate in the men, and one for the same interaction in the women.</p>
<p>Let’s look at the heat maps by themselves, as they are a little bit small.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">draw</span>(onze_fit_gen_int, <span class="at">select=</span><span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="gamms_1_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Note the very different colour scales for male and female speakers.</p>
<p>And look at the smooths section of the model summary:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(onze_fit_gen_int)<span class="sc">$</span>s.table</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                 edf   Ref.df       Chi.sq      p-value
s(yob):genderF              3.447366 4.245710 191.16420928 0.0000000000
s(yob):genderM              1.224522 1.412964 360.09732544 0.0000000000
s(speech_rate)              2.541233 3.301794   5.69677023 0.1377318834
ti(yob,speech_rate):genderF 3.946719 5.233997  22.54341618 0.0005519912
ti(yob,speech_rate):genderM 1.000523 1.000995   0.02282337 0.8815218793</code></pre>
</div>
</div>
<p>This version of the model claims a statistically significant interaction between year of birth and speech rate <em>for the female speakers</em>. The rough story, from the heatmap, when combined with the smooth for year of birth for the female speakers, is that the change over time is more extreme for faster speakers and less extreme for slower speakers.<a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a></p>
</section>
</section>
<section id="the-parenthetical-m-mixed-effects" class="level2" data-number="8.3">
<h2 data-number="8.3" class="anchored" data-anchor-id="the-parenthetical-m-mixed-effects"><span class="header-section-number">8.3</span> The Parenthetical ‘M’: Mixed Effects</h2>
<p>We can include quite complex random effects structures in GAMMs. As in the case of generalised linear models, we can add random <em>intercepts</em> and random <em>slopes</em>. GAMMs introduce the additional possibility of random <em>smooths</em>. The purpose of random effects in this context is exactly the same as in linear mixed-effects models: we want to capture dependence between observations.</p>
<section id="random-intercepts" class="level3" data-number="8.3.1">
<h3 data-number="8.3.1" class="anchored" data-anchor-id="random-intercepts"><span class="header-section-number">8.3.1</span> Random Intercepts</h3>
<p>When fitting a random intercept in a linear mixed-effects model, using the <code>lme4</code>package, for instance, we use the syntax <code>(1 | x_1)</code> within the model formula to get random intercepts for each level of the variable <code>x_1</code>. In <code>mgcv</code>, we use the syntax <code>s(x_1, bs="re")</code>. That is, we specify that we want random intercepts by changing the basis of a smooth term.</p>
<p>In corpus phonetics, we usually want a random intercept for each speaker and each word. In our previous models, we took mean values for each speaker so that we had a single value for each. This is one strategy to avoid the need for random effects, but it is better to include all the data points we have while building the dependence structure between the in to the model.</p>
<p>We create a data set containing all of the <span class="smallcaps">dress</span> F1 readings, again applying Lobanov 2.0 normalisation.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>onze_dress <span class="ot">&lt;-</span> onze_vowels_full <span class="sc">|&gt;</span> </span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lobanov_2</span>() <span class="sc">|&gt;</span> </span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>    vowel <span class="sc">==</span> <span class="st">"DRESS"</span></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We now have a column with speaker identifiers (<code>speaker</code>) and word identifiers (<code>word</code>). We fit the same structure as the previous model, but with random intercepts for these two variables. Since we are significantly increasing the number of tokens (from 481 to 69925), we should also switch to the <code>bam</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>onze_ri_fit <span class="ot">&lt;-</span> <span class="fu">bam</span>(</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">formula =</span> F1_lob2 <span class="sc">~</span> </span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">s</span>(yob, <span class="at">k =</span> <span class="dv">10</span>, <span class="at">by=</span>gender, <span class="at">bs =</span> <span class="st">"tp"</span>) <span class="sc">+</span></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">s</span>(speech_rate, <span class="at">k =</span> <span class="dv">10</span>, <span class="at">bs =</span> <span class="st">"tp"</span>) <span class="sc">+</span></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ti</span>(yob, speech_rate, <span class="at">by=</span>gender, <span class="at">k=</span><span class="dv">5</span>, <span class="at">bs =</span> <span class="st">"tp"</span>) <span class="sc">+</span></span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">s</span>(speaker, <span class="at">bs =</span> <span class="st">"re"</span>) <span class="sc">+</span> <span class="fu">s</span>(word, <span class="at">bs =</span> <span class="st">"re"</span>),</span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> onze_dress,</span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">family =</span> <span class="fu">scat</span>(<span class="at">link=</span><span class="st">"identity"</span>),</span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">discrete =</span> <span class="cn">TRUE</span>,</span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">nthreads =</span> <span class="fu">detectCores</span>() <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="random-smooths" class="level3" data-number="8.3.2">
<h3 data-number="8.3.2" class="anchored" data-anchor-id="random-smooths"><span class="header-section-number">8.3.2</span> Random Smooths</h3>
<p>We’ll now switch to a data set which is more useful for illustrating random smooths and which will also be useful when we turn to the question of auto-correlation later (<a href="#sec-auto" class="quarto-xref"><span>Section 8.4</span></a>).</p>
<p>The new data set comes from <span class="citation" data-cites="soskuthyHorizontalDiphthongShift2019">Sóskuthy, Hay, and Brand (<a href="references.html#ref-soskuthyHorizontalDiphthongShift2019" role="doc-biblioref">2019</a>)</span>. This paper is an exploratory study of the dynamics of New Zealand English diphthongs. It contains formant trajectories for the <span class="smallcaps">price</span> and <span class="smallcaps">mouth</span> vowels. We will take a series of formant trajectories from the <span class="smallcaps">price</span> vowel, looking at both F1 and F2 values. The full data set is available on GitHub, <a href="">here</a> <!-- Add link --></p>
</section>
</section>
<section id="sec-auto" class="level2" data-number="8.4">
<h2 data-number="8.4" class="anchored" data-anchor-id="sec-auto"><span class="header-section-number">8.4</span> Auto-correlation</h2>
<p>Auto-correlation is an issue in time series data. Often a measurement at one time point is correlated with measurements at nearby time points.</p>
<p>When talking about random effects in the previous section, we discussed them in terms of information. We can think of auto-correlation in the same way. Measurements from nearby time points do not provide independent information.</p>
</section>
<section id="sec-hyp" class="level2" data-number="8.5">
<h2 data-number="8.5" class="anchored" data-anchor-id="sec-hyp"><span class="header-section-number">8.5</span> Hypothesis Testing</h2>
<p>How do we test a GAMM model…</p>
<section id="visual-methods" class="level3" data-number="8.5.1">
<h3 data-number="8.5.1" class="anchored" data-anchor-id="visual-methods"><span class="header-section-number">8.5.1</span> Visual methods</h3>
</section>
<section id="model-summary" class="level3" data-number="8.5.2">
<h3 data-number="8.5.2" class="anchored" data-anchor-id="model-summary"><span class="header-section-number">8.5.2</span> Model summary</h3>
</section>
<section id="itsadugcompareml." class="level3" data-number="8.5.3">
<h3 data-number="8.5.3" class="anchored" data-anchor-id="itsadugcompareml."><span class="header-section-number">8.5.3</span> <code>itsadug::compareML()</code>.</h3>
</section>
</section>
<section id="reporting-a-gamm" class="level2" data-number="8.6">
<h2 data-number="8.6" class="anchored" data-anchor-id="reporting-a-gamm"><span class="header-section-number">8.6</span> Reporting a GAMM</h2>
<p>It is important to know how to report a GAMM in a paper.</p>
<p>This section presents some (somewhat randomly selected) examples of GAMMs ‘in the wild’. All of these are good enough for publication. After the examples, I’ll give a few opinions about what ought to be included in the body of an article and in supplementary materials.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true" aria-current="page">Renwick et al.&nbsp;(2023)</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false">Brand et al.&nbsp;(2021)</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-3" role="tab" aria-controls="tabset-1-3" aria-selected="false">Derrick and Gick (2021)</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<p><strong>How did you fit the GAMMs?:</strong></p>
<blockquote class="blockquote">
<p>We built a separate GAMM for each combination of allophone, gender, and formant, leading to twenty-eight separate models. The dependent variable was log-means normalized formant values, from five measurement points per token. The models’ independent variables included a nonlinear, continuous smooth term for speaker year of birth (YoB), and a smooth term for measurement point (percent), ordered 20% to 80%. Smooths used four knots. The two smooths were combined into a tensor-product interaction allowing the predicted trajectory to freely vary in shape across YoB. Vowel duration was included as a parametric effect. The random effects structure comprised linear random intercepts and slopes for speaker, word, and collection. Model specifications and summaries appear in Appendix 1. Models were fitted using the mgcv::bam() function in R (Wood, 2017b). <span class="citation" data-cites="renwickBoomerPeakGen2023">(<a href="references.html#ref-renwickBoomerPeakGen2023" role="doc-biblioref">Renwick et al. 2023, 185</a>)</span></p>
</blockquote>
<p>Appendix 1 contains the output of the <code>summary()</code> function for each GAMM model pasted in to a word document. This includes the model formulae.</p>
<p><strong>How did you evaluate the GAMMs?</strong></p>
<blockquote class="blockquote">
<p>We evaluate the GAMMs in two ways: first, via visualizations of predicted measurements, which were extracted from each model. Second, we tested the significance of YoB via model comparison (Renwick &amp; Stanley, 2020; Stanley et al., 2021). For each GAMM, we constructed a model that excluded YoB but was otherwise identical. Each “dropped” model was evaluated against its “full” equivalent via itsadug:: compareML() (van Rij, Wieling, Baayen, &amp; van Rijn, 2017), which returns a score and p-value through chi-squared testing of log-likelihood. If the “full” model including birth year is deemed to be significantly better ( p &lt; 0.05) than the “dropped” version without birth year, we infer that the shape and/or position of that vowel’s trajectory is meaningfully predicted by speaker year of birth. <span class="citation" data-cites="renwickBoomerPeakGen2023">(<a href="references.html#ref-renwickBoomerPeakGen2023" role="doc-biblioref">Renwick et al. 2023, 185</a>)</span></p>
</blockquote>
<p><strong>How did you report the results?</strong></p>
<blockquote class="blockquote">
<p>Model comparisons confirmed that for all models save one (F2 of PRY, for women) the inclusion of YoB provided a significant improvement in fit over an identical GAMM lacking that term. For all allophones (including women’s <span class="smallcaps">pry</span>, whose F1 model is improved by YoB), there is significant variation in trajectory shape and/or vowel space position across time. <span class="citation" data-cites="renwickBoomerPeakGen2023">(<a href="references.html#ref-renwickBoomerPeakGen2023" role="doc-biblioref">Renwick et al. 2023, 186</a>)</span></p>
</blockquote>
<p>A footnote indicates that <span class="smallcaps">pry</span> has the lowest token count and the prediction that increased tokens would lead to an improvement by adding their year of birth term.</p>
<p>A series of figures are then described in detail. For instance:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/gamms/renwick_etal_fig5.png" class="img-fluid figure-img"></p>
<figcaption>Figure 5 from Renwick et al.&nbsp;2023, p.&nbsp;189.</figcaption>
</figure>
</div>
<p>The description is as follows:</p>
<blockquote class="blockquote">
<p>Young Georgians’ trajectories are modeled in Figure 5, including birth years 1974 (left) and 2000 (right), representing Gen X and Gen Z. The vowel systems shown here are markedly different from older speakers’. For all groups, the /aɪ/ allophones’ offset has a lower F1 than <span class="smallcaps">bat</span>, indicating its trajectory lengthening. <span class="smallcaps">bought</span> remains slightly backer than <span class="smallcaps">bot</span>, although they have similar ingliding trajectories. The lax front vowels have shifted. <span class="smallcaps">bat</span> is ingliding and retracted, while <span class="smallcaps">bet</span> is lower in Gen Z compared to Gen X. <span class="smallcaps">bait</span> is very peripheral, especially in Gen Z, compared to the other front vowels, with a lower F1 and higher F2 throughout its trajectory. <span class="smallcaps">bait</span> and <span class="smallcaps">bet</span> are very distinct, and their trajectories do not cross; there is no <span class="smallcaps">face</span>/<span class="smallcaps">dress</span> swapping for these speakers. These effects are stronger for Gen Z than for Gen X. The picture shown for late twentieth century Georgians is not the SVS; instead, the Gen X speakers show a retreat from the SVS, characterized by retracting front lax vowels, which is consistent with the LBMS among younger, Gen Z speakers. [<span class="citation" data-cites="renwickBoomerPeakGen2023">Renwick et al. (<a href="references.html#ref-renwickBoomerPeakGen2023" role="doc-biblioref">2023</a>)</span>, 187</p>
</blockquote>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<p>As in the <span class="citation" data-cites="renwickBoomerPeakGen2023">Renwick et al. (<a href="references.html#ref-renwickBoomerPeakGen2023" role="doc-biblioref">2023</a>)</span> example, <span class="citation" data-cites="brandSystematicCovariationMonophthongs2021">Brand et al. (<a href="references.html#ref-brandSystematicCovariationMonophthongs2021" role="doc-biblioref">2021</a>)</span> reports <em>many</em> GAMMs fit to readings from vowels. There is significantly more detail concerning the GAMMs in supplementary material for this paper hosted on <code>osf.io</code> and <code>github.com</code>. There is <em>less</em> detail in the paper itself, because the GAMMs are one stage in a larger method.</p>
<p><strong>How did you fit the GAMMs?</strong></p>
<p>In the paper the use of GAMMs is described as follows in the methodology section:</p>
<blockquote class="blockquote">
<p>Speaker intercepts from linear mixed effects regression models have been used to index speaker advancement in sound change (Drager &amp; Hay, 2012; Sóskuthy et al., 2017). The data we analyse here contains some non-linear effects, and thus an appropriate modelling technique to obtain the speaker intercepts from our data is generalised additive mixed modelling (GAMMs). This is because we want to model the normalised F1 and F2 of each of the 10 monophthongs separately, whilst also reliably capturing non-linear changes across time (see Winter &amp; Wieling, 2016; Sóskuthy, 2017; Wieling, 2018 &amp; Chuang et al., Chuang, Fon, &amp; Baayen, 2020 for introductory tutorials on GAMMs). We fitted separate models to normalised F1 and F2 for each of the 10 vowels in the data set, giving 20 models in total.</p>
<p>All models were fitted using the same formula via the mgcv package in R (Wood, 2017), with fixed-effects comprising separate smooths over year of birth for each gender (using an adaptive smooth basis with 10 knots), as well as a smooth over speech rate. Speech rate is calculated as syllables per second for the transcript (where each speaker’s recording is spread across several transcripts. Individual transcripts contain on average approximately 6 min of speech). Random intercepts were included for speaker and word form. We then combined all of the speaker intercepts from the separate models, providing us with a final 481 X 20 data set, where each of the 481 speakers had a separate intercept for each of the 20 vocalic variables. <span class="citation" data-cites="brandSystematicCovariationMonophthongs2021">(<a href="references.html#ref-brandSystematicCovariationMonophthongs2021" role="doc-biblioref">Brand et al. 2021, 8</a>)</span></p>
</blockquote>
<p><strong>How did you report the results?</strong></p>
<p>In the results section, the GAMMs are plotted, again in a vowel space: <img src="images/gamms/brand21.png" class="img-fluid" alt="Figure 5 from Brand et al.&nbsp;(2021)."></p>
<p>The GAMMs are discussed as follows:</p>
<blockquote class="blockquote">
<p>While our primary purpose in fitting the GAMMs is to control for speaker factors such as year of birth, the major patterns revealed by the GAMMs also facilitate an understanding of sound change occurring over this time period, which is a necessary precursor to interpreting the patterns of co-variation. Fig. 5 shows the trajectories of vowel change, based on GAMM smooths over year of birth. All vowels have undergone some change, either linear or non-linear, based on the year of birth smooth effect (all p-values &lt; 0.001), see the Supplementary Materials for the model summaries. An animated and interactive version of the changes is available in the Shiny application. <span class="citation" data-cites="brandSystematicCovariationMonophthongs2021">(<a href="references.html#ref-brandSystematicCovariationMonophthongs2021" role="doc-biblioref">Brand et al. 2021, 9</a>)</span></p>
</blockquote>
<p>The above notes statistical significance of the year of birth smooths and points readers to mode detail if required. The next paragraphs of the results section describe the patterns in the plot in light of existing literature.</p>
<p><strong>What is in supplementary material?</strong></p>
<p>The supplementary material for the analysis is hosted on GitHub. It contains two relevant sections. First, a section with the code for <em>fitting</em> the models, which includes discussion of the model formula and the R code used to fit a series of models with the same formula (<a href="https://nzilbb.github.io/Covariation_monophthongs_NZE/Covariation_monophthongs_analysis.html#GAMM_modelling">here</a>). Second, there is a section containing all of the model summaries (<a href="https://nzilbb.github.io/Covariation_monophthongs_NZE/Covariation_monophthongs_analysis.html#Sound_change_GAMM_summaries">here</a>). These summaries do not contain information about random effects. We have seen above that this radically reduces the amount of time required to generate a model summary.</p>
<p>In addition, an online interactive is provided for exploring the analysis in the paper (including the GAMMs). See <a href="https://onze.shinyapps.io/Covariation_shiny/">here</a>.</p>
</div>
<div id="tabset-1-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-3-tab">
<p><span class="citation" data-cites="derrickGaitChangeTongue2021">Derrick and Gick (<a href="references.html#ref-derrickGaitChangeTongue2021" role="doc-biblioref">2021</a>)</span> is a shorter piece for <em>Scientific Reports</em>. Space is at a premium so a short explanation is given.</p>
<p><strong>How did you fit the GAMMs</strong>?</p>
<p>The discussion of the GAMMs in the methodology sections comes after an earlier set of steps have already been described. Justification is given in terms of non-linearity in the data.</p>
<blockquote class="blockquote">
<p>Next, and in order to make sense of this highly non-linear data, we ran a generalized additive mixed-effects model (GAMM) on the data shown in Eq. (7). GAMMs are extremely effective for the analysis of non-linear data, and are therefore highly suitable for the analysis of the critical fluctuations captured in Eq. (6).</p>
<p>Equation (7), written in R-code, describes a generalized additive mixed-effects model, comparing Fluctuation based on tongue-front displacement (TFd) and the fluctuation time slice position (FTS), forming a 3-dimensional tensor (te) field [te(FTS, TFd)]. The random effects factor out participant variability in a 3-dimensional tensor field [s(FTS, TFd, Participant, bs = “fs”, m = 1)], as well as random-effect smooths for syllables per second [s(SPS, Participant, bs = “re”)], and token type [s(Token type, Participant, bs = “re”)]. In order to correct for autocorrelation effects, we ran the GAMM, calculated an estimate for a start value ρ from that first run, and provided that ρ to a second run of the GAMM model, along with an indicator identifying the first position of our time slices. This removes most of the autocorrelation, maximizing the accuracy of the resulting statistical model output.</p>
<p>Equation (7) produces an output that shows the relationship between critical fluctuation, token position, and tongue-front displacement range, highlighting regions of significant difference. And with these methods, we were able to identify whether tongue-front displacement range affected speech-rate range, and whether tongue-front displacement range had any influence on the timing slice positions of critical fluctuations. <span class="citation" data-cites="derrickGaitChangeTongue2021">(<a href="references.html#ref-derrickGaitChangeTongue2021" role="doc-biblioref">Derrick and Gick 2021, 8</a>)</span></p>
</blockquote>
<p>Equation 7 is the following R formula:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gam</span>(Fluctuation <span class="sc">~</span> <span class="fu">te</span>(FTS, TFd) <span class="sc">+</span> </span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>      <span class="fu">s</span>(FTS, TFd, Participant, <span class="at">bs =</span> <span class="st">"fs"</span>, <span class="at">m =</span> <span class="dv">1</span>) <span class="sc">+</span> </span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>      <span class="fu">s</span>(SPS, Participant, <span class="at">bs =</span> <span class="st">"re"</span>) <span class="sc">+</span> </span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>      <span class="fu">s</span>(Tokentype, Participant, <span class="at">bs =</span> <span class="st">"re"</span>)</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note, btw, the use of all both random smooths (the second smooth term) and two smooth terms defining random slopes (the third and fourth smooth terms in the model).</p>
<p>The methods sections describes the method used to deal with autocorrelation as well.</p>
<p>Finally, a paragraph is given to explain the relationship between the output of the GAMMs and the actual subject matter of the research.</p>
<p>Citations are given in footnotes to R itself, the <code>mgcv</code> package and the <code>itsadug</code> package.</p>
<p><strong>How did you report the results?</strong></p>
<p>There results section reports both the significance of smooth terms and discusses a plot.</p>
<p>The main plot is Figure 7:</p>
<p><img src="images/gamms/derrick_gick_2021.png" class="img-fluid" alt="Figure 7 from Derrick and Gick (2021)."> The discussion is as follows:</p>
<blockquote class="blockquote">
<p>Generalized additive mixed-effects model analysis of critical fluctuations during the time course of token production by tongue-front displacement range are shown in Fig. 7. The model shows that speakers producing tokens with the lowest tongue-front displacement ranges have relatively higher critical fluctuations in the early part of their token productions, spanning from the first vowel through the first flap into the middle of the second vowel. In contrast, they show much lower rates of critical fluctuation from the second half of the second vowel, through the second flap to the end of the third vowel. This constitutes evidence of end-state comfort effects for speakers producing token types with narrow tongue-front displacement ranges.</p>
<p>For speakers producing token types in the middle of the group, there are no end-state comfort effects, but instead the most effort made during the second flap. For speakers with very wide tongue-front displacement ranges, there is again statistically significant evidence for end-state comfort effects, with extra beginning-state effort for the initial vowel. These are the same speakers producing token types that demonstrate two categorically different patterns of motion—one for slow speech, and one for fast speech.</p>
<p>Figure 7 shows the regions of significance for the GAMM whose model output is shown in Table 3. These results show that all of the model parameters are significant, and most importantly that the tensor field shown in Fig. 7 accounts for a significant portion of the variance of the data. This includes the fixed-effect tensor relating tongue-front displacement range and critical fluctuation along time slices, as well as the random-effects for participant, token type, and reiterant speech rate. The entire GAMM accounts for an adjusted r2 of 0.452, explaining 47% of the deviance in critical fluctuations in this dataset. <span class="citation" data-cites="derrickGaitChangeTongue2021">(<a href="references.html#ref-derrickGaitChangeTongue2021" role="doc-biblioref">Derrick and Gick 2021, 10–11</a>)</span></p>
</blockquote>
<p>Significance information is given as Table 3.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/gamms/derrick_gick_2021_t3.png" class="img-fluid figure-img"></p>
<figcaption>Table 3 from Derrick and Gick (2021).</figcaption>
</figure>
</div>
<p><strong>What is in supplementary materials?</strong></p>
<p>There is a repository on OSF (<a href="https://osf.io/7k4ja/?view_only=83ab3adeb363481795ba97e2b481cd9e">here</a>). Code is available as an Rmarkdown or as an HTML file to be downloaded and read in the browser.</p>
<p>The analysis shows that the p-values in Table 3 are generated by using <code>itsadug::compareML()</code> and fitting the models with maximum likelihood estimation. With more space, this might have been discussed in the paper. But there is not always more space!</p>
</div>
</div>
</div>
<p>What can we take from the above examples? Here are some reflections:</p>
<ul>
<li>You don’t need to say <em>everything</em> in the paper. You can use supplementary material. This should include, at a minimum, the specific code you used to fit the model and generate any reported p-values (not <em>just</em> the summary).
<ul>
<li>This can be important for seeing, e.g., whether the ‘difference smooth’ structure was used (see discussion above).</li>
</ul></li>
<li>Cite the specific packages you used. The function <code>citation()</code> in R can be very helpful for this (e.g.&nbsp;look at the results of <code>citation('mgcv')</code>). I often use the package <code>grateful</code> to generate a bibliography of all of the packages I use within an analysis project.</li>
<li>Visualisations are an important part of evaluating GAMMs (moreso than for linear models where a single coefficient is what you are testing).
<ul>
<li>Your discussion section might consist of description of the specific trajectories you find visually.</li>
</ul></li>
<li>Supplementary material opens up many windows (including e.g., the interactive application in <span class="citation" data-cites="brandSystematicCovariationMonophthongs2021">Brand et al. (<a href="references.html#ref-brandSystematicCovariationMonophthongs2021" role="doc-biblioref">2021</a>)</span>). This is also, however, a potential time sink. Try to balance the costs and benefits when you go beyond the requirements of open science and into the realm of web development!</li>
</ul>
</section>
<section id="sec-resources" class="level2" data-number="8.7">
<h2 data-number="8.7" class="anchored" data-anchor-id="sec-resources"><span class="header-section-number">8.7</span> Further Resources</h2>
<ul>
<li>Márton Sóskuthy’s GAMM tutorial: https://arxiv.org/pdf/1703.05339.pdf</li>
<li>Márton Sóskuthy’s paper compared multiple significance testing strategies for error rates: https://www.sciencedirect.com/science/article/pii/S009544702030108X#s0070</li>
<li>Martijn Wieling’s tutorial: https://www.sciencedirect.com/science/article/pii/S0095447017301377</li>
</ul>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Resources have also been written by non-Martins. I will add some soon!</p>
</div>
</div>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list" style="display: none">
<div id="ref-brandSystematicCovariationMonophthongs2021" class="csl-entry" role="listitem">
Brand, James, Jen Hay, Lynn Clark, Kevin Watson, and Márton Sóskuthy. 2021. <span>“Systematic Co-Variation of Monophthongs Across Speakers of <span>New Zealand English</span>.”</span> <em>Journal of Phonetics</em> 88: 101096. <a href="https://www.sciencedirect.com/science/article/pii/S0095447021000711">https://www.sciencedirect.com/science/article/pii/S0095447021000711</a>.
</div>
<div id="ref-derrickGaitChangeTongue2021" class="csl-entry" role="listitem">
Derrick, Donald, and Bryan Gick. 2021. <span>“Gait Change in Tongue Movement.”</span> <em>Scientific Reports</em> 11 (1): 16565. <a href="https://doi.org/10.1038/s41598-021-96139-4">https://doi.org/10.1038/s41598-021-96139-4</a>.
</div>
<div id="ref-renwickBoomerPeakGen2023" class="csl-entry" role="listitem">
Renwick, Margaret E. L., Joseph A. Stanley, Jon Forrest, and Lelia Glass. 2023. <span>“Boomer <span>Peak</span> or <span>Gen X Cliff</span>? <span>From SVS</span> to <span>LBMS</span> in <span>Georgia English</span>.”</span> <em>Language Variation and Change</em> 35 (2): 175–97. <a href="https://doi.org/10.1017/S095439452300011X">https://doi.org/10.1017/S095439452300011X</a>.
</div>
<div id="ref-soskuthyGeneralisedAdditiveMixed2017" class="csl-entry" role="listitem">
Sóskuthy, Márton. 2017. <span>“Generalised Additive Mixed Models for Dynamic Analysis in Linguistics: A Practical Introduction.”</span> arXiv.org. March 15, 2017. <a href="https://arxiv.org/abs/1703.05339v1">https://arxiv.org/abs/1703.05339v1</a>.
</div>
<div id="ref-soskuthyHorizontalDiphthongShift2019" class="csl-entry" role="listitem">
Sóskuthy, Márton, J. Hay, and James Brand. 2019. <span>“Horizontal Diphthong Shift in <span>New Zealand English</span>.”</span> In. <a href="https://www.semanticscholar.org/paper/HORIZONTAL-DIPHTHONG-SHIFT-IN-NEW-ZEALAND-ENGLISH-S%C3%B3skuthy-Hay/cd1bd700686b3d1270be5536e5881e8946ba57ab">https://www.semanticscholar.org/paper/HORIZONTAL-DIPHTHONG-SHIFT-IN-NEW-ZEALAND-ENGLISH-S%C3%B3skuthy-Hay/cd1bd700686b3d1270be5536e5881e8946ba57ab</a>.
</div>
<div id="ref-wielingAnalyzingDynamicPhonetic2018" class="csl-entry" role="listitem">
Wieling, M. 2018. <span>“Analyzing Dynamic Phonetic Data Using Generalized Additive Mixed Modeling: <span>A</span> Tutorial Focusing on Articulatory Differences Between <span>L1</span> and <span>L2</span> Speakers of <span>English</span>.”</span> <em>Journal of Phonetics</em> 70: 86–116. <a href="https://doi.org/10.1016/j.wocn.2018.03.002">https://doi.org/10.1016/j.wocn.2018.03.002</a>.
</div>
<div id="ref-wilsonblackOverlookedEffectAmplitude2023" class="csl-entry" role="listitem">
Wilson Black, Joshua, Jennifer Hay, Lynn Clark, and James Brand. 2023. <span>“The Overlooked Effect of Amplitude on Within-Speaker Vowel Variation.”</span> <em>Linguistics Vanguard</em> 9 (1): 173–89. <a href="https://doi.org/10.1515/lingvan-2022-0086">https://doi.org/10.1515/lingvan-2022-0086</a>.
</div>
</div>
</section>
<section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes">
<hr>
<ol>
<li id="fn1"><p> If you are not familiar with this bit of linguistics, just think of this a set of readings derived from a single action. It could just as easily be a set of readings from a brain scan, or a moving slider, or whatever you like. The only restriction is that we have a set of measurements across time of the same event or action.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>We will look at how to do this later in the series.<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p>Hypothesis testing will be covered in more detail in <a href="#sec-hyp" class="quarto-xref"><span>Section 8.5</span></a>.<a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn4"><p> You may have noticed that there is a correlation between speech rate and year of birth. This should be kept in mind when interpreting the model.<a href="#fnref4" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section>

</main> <!-- /main -->
<script>

/* update total correct if #webex-total_correct exists */
update_total_correct = function() {
  console.log("webex: update total_correct");

  var t = document.getElementsByClassName("webex-total_correct");
  for (var i = 0; i < t.length; i++) {
    p = t[i].parentElement;
    var correct = p.getElementsByClassName("webex-correct").length;
    var solvemes = p.getElementsByClassName("webex-solveme").length;
    var radiogroups = p.getElementsByClassName("webex-radiogroup").length;
    var selects = p.getElementsByClassName("webex-select").length;

    t[i].innerHTML = correct + " of " + (solvemes + radiogroups + selects) + " correct";
  }
}

/* webex-solution button toggling function */
b_func = function() {
  console.log("webex: toggle hide");

  var cl = this.parentElement.classList;
  if (cl.contains('open')) {
    cl.remove("open");
  } else {
    cl.add("open");
  }
}

/* check answers */
check_func = function() {
  console.log("webex: check answers");

  var cl = this.parentElement.classList;
  if (cl.contains('unchecked')) {
    cl.remove("unchecked");
    this.innerHTML = "Hide Answers";
  } else {
    cl.add("unchecked");
    this.innerHTML = "Show Answers";
  }
}

/* function for checking solveme answers */
solveme_func = function(e) {
  console.log("webex: check solveme");

  var real_answers = JSON.parse(this.dataset.answer);
  var my_answer = this.value;
  var cl = this.classList;
  if (cl.contains("ignorecase")) {
    my_answer = my_answer.toLowerCase();
  }
  if (cl.contains("nospaces")) {
    my_answer = my_answer.replace(/ /g, "")
  }

  if (my_answer == "") {
    cl.remove("webex-correct");
    cl.remove("webex-incorrect");
  } else if (real_answers.includes(my_answer)) {
    cl.add("webex-correct");
    cl.remove("webex-incorrect");
  } else {
    cl.add("webex-incorrect");
    cl.remove("webex-correct");
  }

  // match numeric answers within a specified tolerance
  if(this.dataset.tol > 0){
    var tol = JSON.parse(this.dataset.tol);
    var matches = real_answers.map(x => Math.abs(x - my_answer) < tol)
    if (matches.reduce((a, b) => a + b, 0) > 0) {
      cl.add("webex-correct");
    } else {
      cl.remove("webex-correct");
    }
  }

  // added regex bit
  if (cl.contains("regex")){
    answer_regex = RegExp(real_answers.join("|"))
    if (answer_regex.test(my_answer)) {
      cl.add("webex-correct");
    }
  }

  update_total_correct();
}

/* function for checking select answers */
select_func = function(e) {
  console.log("webex: check select");

  var cl = this.classList

  /* add style */
  cl.remove("webex-incorrect");
  cl.remove("webex-correct");
  if (this.value == "answer") {
    cl.add("webex-correct");
  } else if (this.value != "blank") {
    cl.add("webex-incorrect");
  }

  update_total_correct();
}

/* function for checking radiogroups answers */
radiogroups_func = function(e) {
  console.log("webex: check radiogroups");

  var checked_button = document.querySelector('input[name=' + this.id + ']:checked');
  var cl = checked_button.parentElement.classList;
  var labels = checked_button.parentElement.parentElement.children;

  /* get rid of styles */
  for (i = 0; i < labels.length; i++) {
    labels[i].classList.remove("webex-incorrect");
    labels[i].classList.remove("webex-correct");
  }

  /* add style */
  if (checked_button.value == "answer") {
    cl.add("webex-correct");
  } else {
    cl.add("webex-incorrect");
  }

  update_total_correct();
}

window.onload = function() {
  console.log("webex onload");
  /* set up solution buttons */
  var buttons = document.getElementsByTagName("button");

  for (var i = 0; i < buttons.length; i++) {
    if (buttons[i].parentElement.classList.contains('webex-solution')) {
      buttons[i].onclick = b_func;
    }
  }

  var check_sections = document.getElementsByClassName("webex-check");
  console.log("check:", check_sections.length);
  for (var i = 0; i < check_sections.length; i++) {
    check_sections[i].classList.add("unchecked");

    let btn = document.createElement("button");
    btn.innerHTML = "Show Answers";
    btn.classList.add("webex-check-button");
    btn.onclick = check_func;
    check_sections[i].appendChild(btn);

    let spn = document.createElement("span");
    spn.classList.add("webex-total_correct");
    check_sections[i].appendChild(spn);
  }

  /* set up webex-solveme inputs */
  var solveme = document.getElementsByClassName("webex-solveme");

  for (var i = 0; i < solveme.length; i++) {
    /* make sure input boxes don't auto-anything */
    solveme[i].setAttribute("autocomplete","off");
    solveme[i].setAttribute("autocorrect", "off");
    solveme[i].setAttribute("autocapitalize", "off");
    solveme[i].setAttribute("spellcheck", "false");
    solveme[i].value = "";

    /* adjust answer for ignorecase or nospaces */
    var cl = solveme[i].classList;
    var real_answer = solveme[i].dataset.answer;
    if (cl.contains("ignorecase")) {
      real_answer = real_answer.toLowerCase();
    }
    if (cl.contains("nospaces")) {
      real_answer = real_answer.replace(/ /g, "");
    }
    solveme[i].dataset.answer = real_answer;

    /* attach checking function */
    solveme[i].onkeyup = solveme_func;
    solveme[i].onchange = solveme_func;

    solveme[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")
  }

  /* set up radiogroups */
  var radiogroups = document.getElementsByClassName("webex-radiogroup");
  for (var i = 0; i < radiogroups.length; i++) {
    radiogroups[i].onchange = radiogroups_func;
  }

  /* set up selects */
  var selects = document.getElementsByClassName("webex-select");
  for (var i = 0; i < selects.length; i++) {
    selects[i].onchange = select_func;
    selects[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")
  }

  update_total_correct();
}

</script>
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../chapters/qualtrics.html" class="pagination-link" aria-label="Loading Data from Qualtrics">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Loading Data from Qualtrics</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../chapters/references.html" class="pagination-link" aria-label="References">
        <span class="nav-page-text">References</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/nzilbb/statistics_workshops/edit/main/chapters/gamms_1.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/nzilbb/statistics_workshops/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>